README.md
## Deprecation notice

This functionality is now officially supported in Relay using the [`relay-test-utils`](https://relay.dev/docs/en/testing-relay-components) package maintained by the Relay team. It is recommended to use that instead of this package.

# relay-mock-network-layer

Provides a network layer for relay modern that returns schema-correct mock data using `addMockFunctionsToSchema` from [`graphql-tools`](http://dev.apollodata.com/tools/graphql-tools/mocking.html#Default-mock-example);

This is useful for working in an environment like [storybook](https://github.com/storybooks/storybook) where you want to work on your components without hitting a live GraphQL server.

## Usage
```js
import {
	Environment,
	Network,
	RecordSource,
	Store
} from 'relay-runtime';
import getNetworkLayer from 'relay-mock-network-layer';
import schema from './graphql.schema.json';

const network = Network.create(getNetworkLayer({
    schema,
    // pass custom mocks as documented in graphql-tools
    // http://dev.apollodata.com/tools/graphql-tools/mocking.html#Customizing-mocks
    mocks: {
        Geography: () => ({
            id: '2',
            countries: [{
                abbreviation: 'US',
                name: 'United States'
            }, {
                abbreviation: 'UK',
                name: 'United Kingdom'
            }],
                usStates: [{
                abbreviation: 'NY',
                name: 'New York'
            }, {
                abbreviation: 'NJ',
                name: 'New Jersey'
            }]
        }),
        Address: () => ({
            country: 'US',
            city: 'New York',
            state: 'NY',
            zipCode: '10012',
        })
    },

    // See https://www.apollographql.com/docs/graphql-tools/mocking.html#Mocking-interfaces
    preserveResolvers: false,

    // Forward along other options to `makeExecutableSchema`.
    ...schemaDefinitionOptions
}));

// Create an environment using this network:
const store = new Store(new RecordSource());
const environment = new Environment({network, store});

// use environment in <QueryRenderer>

```

### Mocking custom scalar types
If your schema has custom scalar types you'll need to use the `resolvers` option to ensure those types get mocked correctly. Pass this option an object containing a resolve function for each custom scalar.

```js
...
import getNetworkLayer from 'relay-mock-network-layer';
import {GraphQLScalarType} from 'graphql';

...

getNetworkLayer({
    schema,
    mocks: {...},
    resolvers: {
        CustomScalar: new GraphQLScalarType({
            name: 'CustomScalar',
            parseLiteral: () => {},
            parseValue: () => {},
            serialize: () => {}
        }),
        CustomScalar2: ...
    }
});
```


CHANGELOG.md
# Changelog

## [3.0.0]

-   Convert to typescript

## [2.3.0] - 2019-02-26

### Added

-   Performance optimzation: only create executable schema when `getNetworkLayer` is called

## [2.1.0] - 2019-02-26

### Added

-   compile to es5 to work in old browsers

## [2.0.0] - 2018-10-05

## Potentially Breaking

-   Updated to graphql-tools@4.0.2

### Added

-   resolveQueryFromOperation option
-   pass additional options to makeExectuableSchema


index.js
'use strict';

import { makeExecutableSchema, addMockFunctionsToSchema } from 'graphql-tools';
import { graphql, printSchema, buildClientSchema } from 'graphql';
import RelayMockNetworkLayerError from './RelayMockNetworkLayerError';

const defaultSchemaDefinitionOptions = {
    resolverValidationOptions: {
        // Silence error regarding missing Relay `Node` type
        // https://github.com/este/este/issues/1513#issuecomment-385752066
        requireResolversForResolveType: false,
    },
};

/**
 * @param {Object} networkConfig - The configuration for the mock network layer.
 * @param {(string|Object)} networkConfig.schema - If string, the graphql schema SDL. If object, the JSON introspection query.
 * @param {Object} networkConfig.mocks - Mock primative resolvers, passed directly to addMockFunctionsToSchema.
 * @param {Object} networkConfig.resolvers - Default resolvers for a schema.
 * @param {Function} [networkConfig.resolveQueryFromOperation] - If relay operation query text does not exist, used to resolve the query from the operation. Useful for persisted query support.
 */
export default function getNetworkLayer({
    schema,
    mocks,
    resolvers,
    resolveQueryFromOperation,

    /** https://www.apollographql.com/docs/graphql-tools/mocking.html#Mocking-interfaces */
    preserveResolvers = false,

    /**
     * https://github.com/apollographql/graphql-tools/blob/master/src/Interfaces.ts#L125
     * @type IExecutableSchemaDefinition
     */
    ...schemaDefinitionOptions
}) {
    schemaDefinitionOptions = {
        ...defaultSchemaDefinitionOptions,
        ...schemaDefinitionOptions,
    };

    if (typeof schema === 'object' && schema.data) {
        schema = printSchema(buildClientSchema(schema.data));
    }

    const executableSchema = makeExecutableSchema({
        typeDefs: schema,
        resolvers,
        ...schemaDefinitionOptions,
    });

    // Add mocks, modifies schema in place
    addMockFunctionsToSchema({ schema: executableSchema, mocks, preserveResolvers });

    return function fetchQuery(operation, variableValues) {
        const query =
            (resolveQueryFromOperation && resolveQueryFromOperation(operation)) || operation.text;

        if (!query) {
            throw new Error(
                'Could not find query, ensure operation.text exists or pass resolveQueryFromOperation.'
            );
        }

        return graphql(executableSchema, query, null, null, variableValues).then(
            // Trigger Relay error in case of GraphQL errors (or errors in mutation response)
            // See https://github.com/facebook/relay/issues/1816

            result => {
                if (result.errors && result.errors.length > 0) {
                    return Promise.reject(new RelayMockNetworkLayerError(result.errors));
                }
                return Promise.resolve(result);
            }
        );
    };
}


src/index.ts
import {
    makeExecutableSchema,
    addMockFunctionsToSchema,
    IExecutableSchemaDefinition,
    IMocks,
} from 'graphql-tools';
import { FetchFunction, RequestParameters } from 'relay-runtime';
import { graphql, printSchema, buildClientSchema } from 'graphql';
import RelayMockNetworkLayerError from './RelayMockNetworkLayerError';

const defaultSchemaDefinitionOptions = {
    resolverValidationOptions: {
        // Silence error regarding missing Relay `Node` type
        // https://github.com/este/este/issues/1513#issuecomment-385752066
        requireResolversForResolveType: false,
    },
};

interface Config extends Omit<IExecutableSchemaDefinition, 'typeDefs'> {
    schema: { data: any } | string;
    mocks?: IMocks;
    resolveQueryFromOperation?: (operation: RequestParameters) => string;
    preserveResolvers?: boolean;
}

/**
 * @param {Object} networkConfig - The configuration for the mock network layer.
 * @param {(string|Object)} networkConfig.schema - If string, the graphql schema SDL. If object, the JSON introspection query.
 * @param {Object} networkConfig.mocks - Mock primative resolvers, passed directly to addMockFunctionsToSchema.
 * @param {Object} networkConfig.resolvers - Default resolvers for a schema.
 * @param {Function} [networkConfig.resolveQueryFromOperation] - If relay operation query text does not exist, used to resolve the query from the operation. Useful for persisted query support.
 */
export default function getNetworkLayer({
    schema,
    mocks,
    resolvers,
    resolveQueryFromOperation,

    /** https://www.apollographql.com/docs/graphql-tools/mocking.html#Mocking-interfaces */
    preserveResolvers = false,

    /**
     * https://github.com/apollographql/graphql-tools/blob/master/src/Interfaces.ts#L125
     * @type IExecutableSchemaDefinition
     */
    ...schemaDefinitionOptions
}: Config): FetchFunction {
    schemaDefinitionOptions = {
        ...defaultSchemaDefinitionOptions,
        ...schemaDefinitionOptions,
    };

    if (typeof schema === 'object' && schema.data) {
        schema = printSchema(buildClientSchema(schema.data));
    }

    const executableSchema = makeExecutableSchema({
        // @ts-ignore
        typeDefs: schema,
        resolvers,
        ...schemaDefinitionOptions,
    });

    // Add mocks, modifies schema in place
    addMockFunctionsToSchema({ schema: executableSchema, mocks, preserveResolvers });

    return async function fetchQuery(request, variables) {
        const query = resolveQueryFromOperation?.(request) || request.text;

        if (!query) {
            throw new Error(
                'Could not find query, ensure operation.text exists or pass resolveQueryFromOperation.'
            );
        }


        const {errors, data} = await graphql(executableSchema, query, null, null, variables);

        // Trigger Relay error in case of GraphQL errors (or errors in mutation response)
        // See https://github.com/facebook/relay/issues/1816
        if (errors && errors.length > 0) {
            throw new RelayMockNetworkLayerError(errors);
        }

        return {data}
    };
}
